trie/inspect_trie.go

```diff
@@ -0,0 +1,279 @@
+	if ownerAddress == "" {
+		table.SetCaption(true, fmt.Sprintf("%v", treeType))
+	} else {
+		table.SetCaption(true, fmt.Sprintf("%v-%v", treeType, ownerAddress))
+	}
+	table.SetAlignment(1)
+	for i := 0; i < len(trieStat.theNodeStatByLevel); i++ {
+		ns := &trieStat.theNodeStatByLevel[i]
+		if ns.IsEmpty() {
+			break
+
+		}
+		table.AppendBulk([][]string{
+			{"-", fmt.Sprintf("%d", i), fmt.Sprintf("%d", ns.ShortNodeCnt.Load()), fmt.Sprintf("%d", ns.FullNodeCnt.Load()), fmt.Sprintf("%d", ns.ValueNodeCnt.Load())},
+		})
+	}
+	table.AppendBulk([][]string{
+		{"Total", "-", fmt.Sprintf("%d", trieStat.totalNodeStat.ShortNodeCnt.Load()), fmt.Sprintf("%d", trieStat.totalNodeStat.FullNodeCnt.Load()), fmt.Sprintf("%d", trieStat.totalNodeStat.ValueNodeCnt.Load())},
+	})
+	table.Render()
+	return sw.builder.String()
+}
+
+// NewInspector return an inspector obj
+func NewInspector(tr *Trie, db *Database, stateRootHash common.Hash, blocknum uint64, jobnum uint64) (*Inspector, error) {
+	if tr == nil {
+		return nil, errors.New("trie is nil")
+	}
+
+	if tr.root == nil {
+		return nil, errors.New("trie root is nil")
+	}
+
+	ins := &Inspector{
+		trie:           tr,
+		db:             db,
+		stateRootHash:  stateRootHash,
+		blocknum:       blocknum,
+		root:           tr.root,
+		result:         make(map[string]*trieTreeStat),
+		totalNum:       (uint64)(0),
+		wg:             sync.WaitGroup{},
+		sem:            semaphore.NewWeighted(int64(jobnum)),
+		eoaAccountNums: 0,
+	}
+
+	return ins, nil
+}
+
+// Run statistics, external call
+func (inspect *Inspector) Run() {
+	accountTrieStat := &trieTreeStat{
+		is_account_trie: true,
+	}
+
+	if _, ok := inspect.result[""]; !ok {
+		inspect.result[""] = accountTrieStat
+	}
+	log.Info("Find Account Trie Tree", "rootHash: ", inspect.trie.Hash().String(), "BlockNum: ", inspect.blocknum)
+
+	inspect.ConcurrentTraversal(inspect.trie, accountTrieStat, inspect.root, 0, []byte{})
+	inspect.wg.Wait()
+}
+
+func (inspect *Inspector) SubConcurrentTraversal(theTrie *Trie, theTrieTreeStat *trieTreeStat, theNode node, height uint32, path []byte) {
+	inspect.ConcurrentTraversal(theTrie, theTrieTreeStat, theNode, height, path)
+	inspect.wg.Done()
+}
+
+func (inspect *Inspector) ConcurrentTraversal(theTrie *Trie, theTrieTreeStat *trieTreeStat, theNode node, height uint32, path []byte) {
+	// print process progress
+	total_num := atomic.AddUint64(&inspect.totalNum, 1)
+	if total_num%100000 == 0 {
+		fmt.Printf("Complete progress: %v, go routines Num: %v\n", total_num, runtime.NumGoroutine())
+	}
+
+	// nil node
+	if theNode == nil {
+		return
+	}
+
+	switch current := (theNode).(type) {
+	case *shortNode:
+		inspect.ConcurrentTraversal(theTrie, theTrieTreeStat, current.Val, height, append(path, current.Key...))
+	case *fullNode:
+		for idx, child := range current.Children {
+			if child == nil {
+				continue
+			}
+			childPath := append(path, byte(idx))
+			if inspect.sem.TryAcquire(1) {
+				inspect.wg.Add(1)
```

Comment:
I haven't used weighted semaphore before, so I had to read up on it. So the idea afaict, is that you have a resource, with `n` units. And at various points, you want to allocate `m`, or `p` units. So one call may be `TryAquire(500)` and another be `TryAcquire(25)`, and this semaphore handles it so we don't over-allocate. 

As far as I can tell, you don't need the advanced semantics of weighted semaphore, because you always want to allocate or free exactly one resource. Maybe there's some simpler way to do it, but as I started writing one based on a counter, I realized that most 'simpler' methods actually becomes more lines of code :)

Reply:
In fact, I also used to write one based on a counter. It's probably an implementation issue, it doesn't seem to control the number of concurrencies very well. sad...

And as you said, the code become less clean and simpler.